Cache and Eviction Strategy: Offline First

CalebKL opened the pull request:
## Related Issue
#21
## Description
Added offline First features. 
fetchWeather function from the repositiory is responsible for fetching weather data. It first checks if there is cached weather data available and if it is still valid. If so, it emits the cached data as a ApiResult.Success result. If not, it makes an API request to fetch the weather data. If the API response is successful, it converts the response to a WeatherEntity, inserts it into the local cache using the weatherDao, and emits the converted data as a ApiResult.Success result.
Finally, in the onCompletion block, it schedules a background worker (UpdateWeatherWorker) using WorkManager to refresh the weather data periodically after 15 Minutes.

## How Can It Be Tested
The Test function passes and emits ApiResult.Success when fetchWeather contains the expected mapped weather data.
## Screenshots (If Applicable)
N/A
## Additional Comments
I also added a notification utility, so when there's a newly refreshed weather update, client gets a notification.
## Checklist

- [ x] New tests were added/Existing Modified

CalebKL committed:
Added Dao, Weather Entity and Database

CalebKL committed:
updated gradle dependency for room and Workmanager

CalebKL committed:
DI for Local module and updated Manifest file to include notification permission

CalebKL committed:
Added Notification Util to update weather info to the user after 15 Min

CalebKL committed:
Updated mappers to include the new entity created

CalebKL committed:
updated fetchWeather function to read data from cache and update after 15 Min using work manager

CalebKL committed:
updated WeatherApp to getWorkManagerConfigs and included Timber for Debugging

CalebKL committed:
Added gson converter gradle dependency and created Converter class to transform hourly and daily weather list

CalebKL committed:
Added typeconverters to DI, Database and updated WeatherEntity Class

CalebKL committed:
Updated mappers and DefaultWeatherRepository with fetching the correct weatherinfo from the cached data

CalebKL committed:
Added testing dependencies

CalebKL committed:
Updated Fakes with the corrected mapped responses

CalebKL committed:
Updated HomeViewModelIntegrationTest

CalebKL committed:
Updated WeatherRepositoryUnitTest

CalebKL committed:
Resolved Conflicts with develop branch

CalebKL commented:
I have ran the tests bitrise is warning about and they are passing. Not sure why but I added Roboelectric test runner and they pass.

jumaallan commented on app/build.gradle.kts:
```diff
@@ -75,7 +75,7 @@ dependencies {
     implementation(libs.kotlinx.serialization)
     implementation(libs.coil)
-
+    implementation(libs.retrofit.converter.gson)
```
use something modern like moshi or kotlinx.serialization

CalebKL replied:
Okay. On it

jumaallan commented on app/src/main/java/com/github/odaridavid/weatherapp/data/weather/local/converters/Converters.kt:
```diff
@@ -0,0 +1,58 @@
+    private val gson = Gson()
+
+    @TypeConverter
+    fun fromHourlyWeatherEntityList(hourlyList: List<HourlyWeatherEntity>): String {
```
Just curious

Why do you want to save a list of entities as a string - is it not better to basically use relationships?

Also, avoid gson as suggested earlier

Use typeconverters only when converting data types not supported by room natively

CalebKL replied:
My bad. Relationships would be a better approach. Let me make the adjustments and remove gson.

odaridavid commented on app/src/main/java/com/github/odaridavid/weatherapp/data/weather/DefaultWeatherRepository.kt:
```diff
@@ -1,57 +1,85 @@
             is IOException -> R.string.error_connection
             else -> R.string.error_generic
         }
-        logger.logException(throwable)
```
This logger is used to send events to dashboards i.e. crashlytics etc,see `FirebaseLogger`, I would prefer this approach over logging locally in the dev environment with Timber.

CalebKL replied:
Okay. That's fine.

odaridavid commented on app/src/main/java/com/github/odaridavid/weatherapp/data/weather/DefaultWeatherRepository.kt:
```diff
@@ -1,57 +1,85 @@
         }
-        logger.logException(throwable)
         emit(ApiResult.Error(errorMessage))
+    }.onCompletion {
+        val constraints = Constraints.Builder()
+            .setRequiredNetworkType(NetworkType.CONNECTED)
+            .build()
+
+        val refreshWeatherRequest = OneTimeWorkRequestBuilder<UpdateWeatherWorker>()
+            .setConstraints(constraints)
+            .setInitialDelay(15, TimeUnit.MINUTES)
+            .build()
+
+        WorkManager.getInstance(context).enqueue(refreshWeatherRequest)
```
Nice work here👌, but currently it looks like the repository is doing more than intended, you could make use of things like use cases or come up with an approach that abstracts the work manager setup and use.

CalebKL replied:
I can  do the use-cases and make the changes

odaridavid reviewed (commented):
Thanks for the contribution 🙌, I'll leave some comments on the go

odaridavid commented on app/src/test/java/com/github/odaridavid/weatherapp/Fakes.kt:
```diff
@@ -2,25 +2,51 @@ package com.github.odaridavid.weatherapp
+            id = 0,
+            main = "fake",
+            description = "fake",
+            icon = "http://openweathermap.org/img/wn/fake@2x.png"
```
Tests are failing because it expects `https` I believe

odaridavid commented on app/src/main/java/com/github/odaridavid/weatherapp/worker/UpdatedWeatherWorker.kt:
```diff
@@ -0,0 +1,55 @@
+                language = language,
+                units = units
+            )
+            notificationUtil.makeNotification("Weather Updated")
```
Refactor to use string resource

odaridavid commented on app/src/main/java/com/github/odaridavid/weatherapp/util/NotificationUtil.kt:
```diff
@@ -0,0 +1,36 @@
+
+    private fun createNotificationChannel() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val name = "notification"
+            val descriptionText = "description"
```
I'd suggest more meaningful names and description on what kind of notificiations this channel provides e.g Weather Updates or something in that line

CalebKL replied:
okay.

CalebKL committed:
Resolved Conflicts with develop branch

CalebKL committed:
Removed Gson and updated type converters

CalebKL committed:
Refactored DAO to use relationships and updated entity and db

CalebKL committed:
updated mappers with the populated weather

CalebKL committed:
Created usecase to sync updated weather and modified default weather repo

CalebKL committed:
removed raw strings and updated strings.xml

CalebKL committed:
removed timber config

CalebKL committed:
Added DI for the WeatherUpdateScheduler

CalebKL committed:
Fixed failing tests and updated Fakes

CalebKL committed:
Merge branch 'develop' into offlineCache

# Conflicts:
#	app/build.gradle.kts
#	gradle/libs.versions.toml

CalebKL committed:
Merge branch 'develop' into offlineCache

# Conflicts:
#	app/build.gradle.kts
#	gradle/libs.versions.toml

CalebKL committed:
Adds Logger to DefaultWeatherRepository, HomeViewModelIntegrationTest and Weather

CalebKL committed:
Adds Logger toHomeViewModelIntegrationTest and WeatherRepositoryUnitTest

CalebKL commented:
@odaridavid I had updated the changes as you had requested. Kindly review. 
@jumaallan I have removed gson and updated the entities to use relations, however, I am not able to populate the hourly and daily lists, I can only view the current weather. Kindly check and the entities. 
There' entities is a bit complex since we have about three nested entities(Weather-Listof(Dailly)-Listof(WeatherResponse). 
The same applies to the hourly. I used type converters to convert the list of weather response then have relations with the weather and daily list/hourly list.

odaridavid commented on app/src/main/java/com/github/odaridavid/weatherapp/data/weather/RefreshWeatherUseCase.kt:
```diff
@@ -0,0 +1,45 @@
+import kotlinx.coroutines.flow.flow
+import javax.inject.Inject
+
+class RefreshWeatherUseCase @Inject constructor(
```
Nice work on the usecase, 

Preferably it would implement an interface i.e. `RefreshWeatherUseCase` that is part of the `core` package to represent business domain use cases, it's also the pattern I intended to use on other parts as functionality increases.

CalebKL replied:
updated. Kindly check

odaridavid commented on app/src/main/java/com/github/odaridavid/weatherapp/di/UseCaseModule.kt:
```diff
@@ -0,0 +1,29 @@
+            notificationUtil = notificationUtil,
+            context = context
+        )
+    }
```
Based on the previous comment this would also change and we wouldn't have to care about doing concrete provides for use cases in di
`@Binds` should work fine.

CalebKL replied:
updated. Kindly check

odaridavid commented on app/src/main/java/com/github/odaridavid/weatherapp/worker/UpdatedWeatherWorker.kt:
```diff
@@ -0,0 +1,54 @@
+        return try {
+            val defaultLocation = getDefaultLocation()
+            val language = getDefaultLanguage()
+            val units = getDefaultUnits()
```
What do you think about using the settings repo instead for the default location, language, and unit values?

CalebKL replied:
updated. Kindly check

CalebKL committed:
Created RefreshWeatherUseCase Interface

CalebKL committed:
Refactored Usecase to implement the interface created and updated DI

CalebKL committed:
Merge develop

CalebKL committed:
Merge develop

odaridavid reviewed (approved):
Looks fine to me ✅, great work on pushing this.

My final comment would be the UseCase implementation should ideally sit in the core/domain and can be done in such a way that the platform-specific needs are implemented in the presentation layer, but that can be followed up later like so
```
class DefaultRefreshWeatherUseCase @Inject constructor(
    private val weatherRepository: WeatherRepository
) : RefreshWeatherUseCase {
    override operator fun invoke(
        defaultLocation: DefaultLocation,
        language: String,
        units: String,
        onSuccess: (data) -> Unit
        onError: (errorType) -> Unit
    ): Flow<Result<Weather>>  = flow{
        weatherRepository.fetchWeatherData(
            defaultLocation = defaultLocation,
            language = language,
            units =units
        ).collect{ result->
            when (result) {
                is Result.Success -> {
                     onSuccess(result.data)
                }
                is Result.Error -> {
                    onError(result.errorType)
                }
            }
        }

    }
}
```

odaridavid committed:
Merge branch 'develop' into offlineCache

odaridavid commented:
<img src="https://github.com/odaridavid/WeatherApp/assets/20339228/52bfbf32-d8fe-474c-bdee-cd9ede12f0cc" width="300px"/>

There seems to be an issue when I run the app, I'll hold off on the merge for now

odaridavid reviewed (changes requested):
What I see on the log cat 
<img width="1440" alt="Screenshot 2023-06-10 at 12 20 19" src="https://github.com/odaridavid/WeatherApp/assets/20339228/cd45fdfe-5658-407f-a42e-134e2f72eca1">

CalebKL commented:
@odaridavid yes. There's a problem with the room relations. I am waiting for @jumaallan to help me out.
Alternatively, can I use Typeconverters for this PR and then work on the room relations in a separate PR?
what do you think?

CalebKL committed:
created CustomWorkerFactory

CalebKL committed:
Updated daos, database and entity

CalebKL committed:
Merge remote-tracking branch 'origin/offlineCache' into offlineCache

CalebKL committed:
updated viewmodel component with singleton component

CalebKL committed:
Added latitude and longitude into the weather core model and updated mappers

CalebKL committed:
refactored weather update scheduler

CalebKL committed:
deleted converters and replaced with room relations

CalebKL committed:
refactored usecase

CalebKL committed:
removed type converters from the local module

CalebKL committed:
refactored repository to read daily and hourly lists

CalebKL committed:
updated weather app with custom worker instead of hiltworker

CalebKL committed:
Maoppers update

CalebKL committed:
Updated Fakes with the Populated Weather

CalebKL committed:
Fixed failing tests

CalebKL commented:
@odaridavid I fixed the error on Workmanager. Also I had to add latitude and longitude to the weather models so I can use it on my entity. 
I will explain why? 
The Populated weather would have consisted of Current, Hourly list and daily list. But since the current weather has a list of weather info responses, we would need to define another entity(Weather Entity) that would have the latitude and longitude parameters. The ID(weatherId) on this entity would then need to the parent entity on the PopulatedWeather as I have done. 
Here is the resource I used.
https://developer.android.com/training/data-storage/room/relationships
(At the bottom there's the implementation on defining nested relationships). 

This also fixes the review @jumaallan had initially requested and removed the type converters. 
Have a look and let me know if there's anything I should update on this PR.
